qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -418,151 +449,227 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
-
-            if not modified:
-                continue
-
-            if isinstance(decomposition, QuantumCircuit):
-                dag.substitute_node_with_dag(
-                    node, circuit_to_dag(decomposition, copy_operations=False)
+        qubits = tuple(dag.find_bit(q).index for q in dag.qubits)
+        if self.qubits_initially_zero:
+            clean, dirty = set(qubits), set()
+        else:
+            clean, dirty = set(), set(qubits)
+
+        tracker = QubitTracker(qubits=qubits, clean=clean, dirty=dirty)
+        return self._run(dag, tracker)
+
+    def _run(self, dag: DAGCircuit, tracker: QubitTracker) -> DAGCircuit:
+        # Start by analyzing the nodes in the DAG. This for-loop is a first version of a potentially
+        # more elaborate approach to find good operation/ancilla allocations. It greedily iterates
+        # over the nodes, checking whether we can synthesize them, while keeping track of the
+        # qubit states. It does not trade-off allocations and just gives all available qubits
+        # to the current operation (a "the-first-takes-all" approach).
+        synthesized_nodes = {}
+
+        for node in dag.topological_op_nodes():
+            qubits = tuple(dag.find_bit(q).index for q in node.qargs)
+            synthesized = None
+            used_qubits = None
+
+            # check if synthesis for the operation can be skipped
+            if (
+                dag.has_calibration_for(node)
+                or len(node.qargs) < self._min_qubits
+                or node.is_directive()
+                or self._definitely_skip_node(node, qubits)
+            ):
+                pass
+
+            # next check control flow
+            elif node.is_control_flow():
+                node.op = control_flow.map_blocks(
+                    partial(self._run, tracker=tracker.copy()),
+                    node.op,
+                    # self.run,
+                    # node.op,
                 )
-            elif isinstance(decomposition, DAGCircuit):
-                dag.substitute_node_with_dag(node, decomposition)
-            elif isinstance(decomposition, Operation):
-                dag.substitute_node(node, decomposition)
 
-        return dag
+            # now we are free to synthesize
+            else:
+                # this returns the synthesized operation and the qubits it acts on -- note that this
+                # may be different than the original qubits, since we may use auxiliary qubits
+                synthesized, used_qubits = self._synthesize_operation(node.op, qubits, tracker)
+
+            # if the synthesis changed the operation (i.e. it is not None), store the result
+            # and mark the operation qubits as used
+            if synthesized is not None:
+                synthesized_nodes[node] = (synthesized, used_qubits)
+                tracker.used(qubits)  # assumes that auxiliary are returned in the same state
```

Comment:
Should this be `tracker.used(used_qubits)`? Update: no, this is probably correct, `used_qubits` is the union of the original nodes's qubits + additional ancilla qubits employed by synthesis. If some of these ancillas were in `|0>`, the synthesis had to return them back to `|0>`, and they are still clean.

Reply:
It should be correct as is, because of the explanation you gave 🙂